text/primitive-projections.md

```diff
@@ -0,0 +1,78 @@
+- Title: What model for primitive projections?
+
+----
+
+# Summary
+
+Projections of record fields currently follow different overlapping paths:
+- constructors-based vs destructors-based normal forms
+- `t.(f params)` vs `f params t` concrete representation
+- unfolded `Proj` vs folded `Proj` vs `App` internal representation
+
+This document is an attempt to propose solutions to clarify the picture.
+
+See also [#14640](https://github.com/coq/coq/pull/14640) which tries to set up a context to stabilize a design choice.
+
+## Issue 1: the relative status of `Const`-based projections and `Proj`-based projections
+
+It is common to write `f param t` to mean a compact `Proj` kernel node. It is also useful to have a non compact representation for partial application of a projection or for an unapplied projection. What to do with it?
+
+*Design choice 1*: not a kernel problem: the user (or vernac level) can declare `Definition f' params x := x.(f params)` and use it later as an ordinary definition
+  - Advantages: a simple model
+  - Drawbacks: the name `f'` is different
+
+*Design choice 2*: Whenever a record is defined, a `Const f` node of same name `f` is made available at the same time and `Proj (f,x)` and `App (Const f, params, x)` are considered convertible by default without needing any form of explicit unfolding. Note that the infrastructure to make them convertible basically already exists in coercions, unification, ...
+  - Advantages: the two representations are fully equivalent (up to the compact representation and the loss of the exact form of parameters)
+  - Drawbacks: the code to identify `Const` and `Proj` might be considered intrusive (but at least, it is alreay there, using `Projection.constant` and `expand_projection` machinery)
+
+*Design subchoice 2.1*: When the syntax `f params t` is used, does it bind automatically to the compact representation (as it is done now), and thus reprinted as `t.(f params)`?
+  - Advantages: compatibility with the current usages
+  - Drawbacks: it does some transformation under the carpet
+
+*Design subchoice 2.2*: Any application of `Const f` to enough arguments is automatically turned by `mkApp` into a `Proj` (this assumes that any constant `f` contains an information on its projection status)? In particular, `Const` is used only for non-applied or not-enough-applied projection.
```

Comment:
FTR, I am strongly opposed to adding more quotients as smart constructors in the kernel. We currently have two of those, namely:
- Automatic concatenation of application nodes.
- Collapse of cast nodes.
The first one is reasonable because it's an isomorphism. The second one less so, because it implicitly erases information from the term.

As a general principle, I believe that quotienting ASTs is a very bad practice. It makes reasoning about them harder, and interferes with higher-level abstractions. We already have to rebuild the abstraction in `EConstr` in order to transparently respect the same invariants in presence of evars, but even that is broken by evar instantiation (i.e. casts might disappear). There are probably other similar issues in the code but we're not aware of them because casts already have a strong tendency to disappear silently.

Doing the same thing with projections would make the situation even worse. Just like casts, the quotient you propose is lossy, so by applying a term you can make information disappear. This can silently hide ill-typed data (e.g. you have broken parameters but since the projection is fully applied they magically disappear) and it would also interfere with evars.

ASTs ought to be what they claim to be, i.e. trees, not semantical objects that have some built-in non-trivial equations. Otherwise, we wouldn't we also include full strong normalization in the smart constructors?

Reply:
About how much the concrete syntax can differ from internal representation, I think that I agree that there should be a mode (i.e. at least with `Set Printing All` extended with universes and existential instances printing) where we can display the internal representation explicitly.

About primitive projections, are you proposing a "design choice" 2.3 which at the time a record with projection name `f` is declared, do the following:
  - provides the ability to use compact `Proj(f,c)` nodes with concrete syntax `c.(f _)` (that is `c.(f)` when parameters are declared implicit)
  - defines a constant `Const(f)` bound to `fun params c => c.(f _)` with concrete notation `f` and with non-compact representation `App(Const(f),[|params;c|])` whose concrete notation is itself `f params c`

About compatibility, does this mean that all usages of `f params c` which currently go to `Proj(f,c)` would then go instead to `App(Const(f),[|params;c|])` hoping that it works almost observationally the same, telling users to use instead `c.(f params)` if they really want the compact representation?

Regarding the compatibility for records not tagged as primitive, is the syntax `c.(f params)` disallowed? Is it kept as an alias for `f params c`? Or is the plan to have a new notation say `c.[f]` for the compact representation of projections in records tagged as primitive so that it does not interfere with the current usages?

Would it mean also keeping all current code which equates `App(Const(f),[|params;c|])` and `Proj(f,c)` in unification, pattern-matching, notations, etc. or the plan would instead to invite users to make a clear and explicit difference between `f params c` and `c.(f _)`?

Reply:
I've already stated it several times around before, but assuming we want to preserve `f.(p)` for arbitrary applications (which is reasonable given that otherwise it'd be a serious backwards incompatible change) we desperately need a faithful way to display primitive projections to the user. The issues people have experienced so far is not so much a change of semantics rather than an utter confusion at things displayed the same (notably, the folded flag and the proj vs app distinction). Therefore, we *must* introduce a new syntax that will unambiguously represent `Proj`.

Reply:
Someone needs to unilaterally decide the new syntax or it will stay in bikeshedding mode.

Reply:
#14640 is incomparable to what I am saying, since part of it is a prerequisite and part of it is unrelated. My proposal is just that we print `Proj` nodes with a unique syntax different from anything we've been using so far. Obviously it requires preserving those nodes as far as possible inside `constr_expr` when detyping, but it has no other effects on the runtime.

Reply:
> My proposal is just that we print `Proj` nodes with a unique syntax different from anything we've been using so far. 

Do you already have in mind a new syntax that could be added to the document? Also what are your plans about existing uses of the `x.(f)` notation in primitive records at parsing time?